Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py

```diff
@@ -0,0 +1,156 @@
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import hashlib
+import hmac
+import json
+
+from flask import current_app, json as fjson, request
+from reporelaypy import Database
+from webkitflaskpy import AuthedBlueprint
+from webkitcorepy import Environment, string_utils
+
+TIMEOUT = 30 * 60
+
+
+class HookProcessor(object):
+    INBOUND_KEY = 'inbound-hooks'
+    WORKER_HOOKS = 'worker-hooks'
+    TYPES = ('push',)
+
+    @classmethod
+    def is_valid(cls, type, data):
+        if type not in cls.TYPES or not isinstance(data, dict):
+            return False
+        if type == 'push' and data.get('ref'):
+            return True
+        return False
+
+    def __init__(self, checkout, database=None, num_workers=1, worker_index=0, callbacks=None):
+        self.checkout = checkout
+        self.database = database or Database()
+        self.num_workers = num_workers
+        self.worker_index = worker_index
+        self.callbacks = callbacks or dict()
+
+    def process_hook(self, type, data):
+        if not self.is_valid(type, data):
+            return None
+
+        self.callbacks.get(type, lambda _: _)(data)
+
+    def process_worker_hook(self, type, data):
+        if not self.is_valid(type, data):
+            return None
+
+        branch = data.get('ref')
+        if type == 'push' and branch:
+            try:
+                if branch.startswith('refs/heads/'):
+                    branch = branch[len('refs/heads/'):]
+                self.checkout.update_for(branch, track=True)
+            except BaseException as e:
+                sys.stderr.write('{}\n'.format(e))
+
+    def process_hooks(self):
+        with self.database.lock(name='lock_{}'.format(self.INBOUND_KEY), timeout=60):
+            for key in self.database.scan_iter('{}:*'.format(self.INBOUND_KEY)):
+                data = json.loads(self.database.get(string_utils.decode(key)))
+                self.process_hook(type=data.get('type'), data=data.get('data'))
+
+                encoded = json.dumps(data)
+                for i in range(self.num_workers):
+                    digest = hashlib.md5()
+                    digest.update(string_utils.encode(encoded))
```

Comment:
Do we need to calculate md5 for encoded multiple times? we can just calculate hexdigest once out of this for loop.
Also worth mentioning md5 is not safe, but might be OK for getting a unique key.

Reply:
Just looking for a unique key.